### PR TITLE
[FEATURE] Update diagnostic checklist to do linting checks

### PR DESCRIPTION
**Pull request by kenwade4**
Changes proposed in this pull request:
- Update diagnostic checklist to do linting checks
- Make linting check be a part of experimental checks instead of production checks
- Also move camel_to_snake func to great_expectations/util.py

### Definition of Done

- [x] My code follows the Great Expectations [style guide](https://docs.greatexpectations.io/docs/contributing/style_guides/code_style)
- [x] I have performed a [self-review](https://docs.greatexpectations.io/docs/contributing/contributing_checklist) of my own code
- [x] I have run any local integration tests and made sure that nothing is broken.

Thank you for submitting!

### CODE REVIEW

File: great_expectations/core/expectation_diagnostics/expectation_diagnostics.py [hunk 1]
```diff
@@ -1,3 +1,6 @@
+import inspect
+import os.path
```
Comment (cdkini) [inline great_expectations/core/expectation_diagnostics/expectation_diagnostics.py]: Nitpick - could we just import `os` to be consistent with the rest of the codebase?

File: great_expectations/expectations/expectation.py [hunk 2]
```diff
@@ -1546,6 +1536,7 @@ def _get_maturity_checklist(
                 tests
             )
         )
+        experimental_checks.append(ExpectationDiagnostics._check_linting(self))
```
Comment (cdkini) [inline great_expectations/expectations/expectation.py:1539]: Should `check_linting` be a public method? I feel little iffy invoking a private method if we can avoid it.

File: great_expectations/util.py [hunk 3]
```diff
@@ -95,6 +95,9 @@
     "rendered_data_docs": "rendered_data_doc",
 }
 
+p1 = re.compile(r"(.)([A-Z][a-z]+)")
+p2 = re.compile(r"([a-z0-9])([A-Z])")
```
Comment (cdkini) [inline great_expectations/util.py:99]: Could we give these more descriptive names? What do `p1` and `p2` do here?

File: great_expectations/util.py [hunk 4]
```diff
@@ -122,6 +125,11 @@ def singularize(plural_ge_noun):
         )
 
 
+def camel_to_snake(name):
+    name = p1.sub(r"\1_\2", name)
+    return p2.sub(r"\1_\2", name).lower()
```
Comment (cdkini) [inline great_expectations/util.py:130]: Could we type annotate the args and return here? Also, could we add a few unit tests? I trust your regex skills a lot more than I do mine but I think a few tests to gain coverage would be great!

File: great_expectations/core/expectation_diagnostics/expectation_diagnostics.py [hunk 5]
```diff
@@ -378,10 +395,73 @@ def _check_renderer_methods(
     @staticmethod
     def _check_linting(expectation_instance) -> ExpectationDiagnosticCheckMessage:
         """Check if linting checks pass for Expectation"""
-        # TODO: Perform linting checks instead of just giving thumbs down
+        sub_messages = []
+        passed = False
+        black_ok = False
+        isort_ok = False
+        file_and_class_names_ok = False
+
```
Comment (cdkini) [inline great_expectations/core/expectation_diagnostics/expectation_diagnostics.py]: Could we add type annotations to all variable declarations throughout?
```suggestion
        sub_messages: List[dict] = []
        passed: bool = False
        black_ok: bool = False
        isort_ok: bool = False
        file_and_class_names_ok: bool = False
```

File: great_expectations/core/expectation_diagnostics/expectation_diagnostics.py [hunk 6]
```diff
@@ -378,10 +395,73 @@ def _check_renderer_methods(
...
+        expectation_camel_name = rx_expectation_instance_repr.match(
+            repr(expectation_instance)
+        ).group(1)
```
Comment (cdkini) [inline great_expectations/core/expectation_diagnostics/expectation_diagnostics.py]: Will this ever fail? What if group doesn't have 1?

Nitpick - could we split up this logic into separate lines? Perhaps the match on one and the group on the other?

File: great_expectations/core/expectation_diagnostics/expectation_diagnostics.py [hunk 7]
```diff
@@ -378,10 +395,73 @@ def _check_renderer_methods(
...
+            isort_ok = isort.check_code(code, **isort.profiles.black)
+            if not isort_ok:
+                sub_messages.append(
+                    {
+                        "message": "Your code would be reformatted with isort",
+                        "passed": False,
+                    }
+                )
+
+        passed = black_ok and isort_ok and file_and_class_names_ok
```
Comment (cdkini) [inline great_expectations/core/expectation_diagnostics/expectation_diagnostics.py:488]: love it!

File: great_expectations/core/expectation_diagnostics/expectation_diagnostics.py [hunk 8]
```diff
@@ -18,7 +21,21 @@
...
+    import black
+except ImportError:
+    black = None
+
+try:
+    import isort
+except ImportError:
+    isort = None
+
+rx_expectation_instance_repr = re.compile(r"<.*\.([^\.]*) object at .*")
```
Comment (cdkini) [inline great_expectations/core/expectation_diagnostics/expectation_diagnostics.py]: If this regex is only used in the single method, could we perhaps do this within the function? I'd rather clean up the global namespace if at all possible, even if it means being slightly less performant.

File: great_expectations/util.py (hunk 4 above)
Comment (cdkini) [reply great_expectations/util.py:130]: Oops now seeing you just ported this over - will defer to you here.

**Review by cdkini — APPROVED** [review]
LGTM - left comments around type annotations and general style that I'd love resolved but I don't want to block on that.

File: great_expectations/expectations/expectation.py (hunk 2 above)
Comment (kenwade4) [reply great_expectations/expectations/expectation.py:1539]: All the staticmethod/classmethod decorated methods in the ExpectationDiagnostics class start with `_check_`

File: great_expectations/util.py (hunk 3 above)
Comment (kenwade4) [reply great_expectations/util.py:99]: Just moved as-is from great_expectations/expectations/expectation.py

File: great_expectations/util.py (hunk 4 above)
Comment (kenwade4) [reply great_expectations/util.py:130]: Just moved as-is from great_expectations/expectations/expectation.py

**Pull request merged**